Add a callback for increasing a memory cap

oerling opened the pull request:
A MemoryUsageTracker can try to raise its limit instead of throwing an
exception. Child and parent trackers can in principle both have
caps. First we check against the cap of parents, then of children. If
the parent raise succeeds and the child raise fails, the parent raise
stays in effect. The grow callback is called with the tracker set as
if the allocation that exceeded the limit succeeded. The callback may
raise the limit to accommodate the allocation or return false to
indicate that the allocation should be reverted and an exception
thrown.

The practical use case has a Task level tracker with a cap for each
Task.  The grow callback of this will look at other Tasks and see if
growing is possible or if Tasks should spill or be killed.

pedroerp commented on velox/common/memory/MemoryUsageTracker.cpp:
```diff
@@ -53,24 +69,30 @@ void MemoryUsageTracker::updateInternal(UsageType type, int64_t size) {
-      parent_->updateInternal(type, -size);
+      (newUsage > usage(maxMemory_, type) || totalBytes > total(maxMemory_))) {
+    bool limitRaised = false;
+    if (growCallback_) {
```
nit: maybe some more concise like:

```
  if (growCallback_ && !growCallback_(type, size, *this)) {
    .. error handling block ...
  }
```

pedroerp commented on velox/common/memory/MemoryUsageTracker.cpp:
```diff
@@ -53,24 +69,30 @@ void MemoryUsageTracker::updateInternal(UsageType type, int64_t size) {
+    if (!limitRaised) {
+      // Exceeded the limit.  revert the change to current usage.
+      if (parent_) {
+        parent_->updateInternal(type, -size);
```
couldn't this trigger growCallback in the parent as well? Maybe split the functions in an updateInternal() and updateInternalUnsafe() (or similar), the ensure the latter wouldn't do the same checks and potentially trigger the callback?

pedroerp reviewed (commented):
Code looks great, just a few comments:

oerling commented:
Thank you for the review. Replies following. 

 

+    std::lock_guard<std::mutex> l(mutex_);

is acquiring a lock in an exception handling code safe? Can't this lock guard throw as well and end up aborting the process? Maybe be safer to just a acquire the std::exception_ptr and handle and re-throw outside of the exception handling code.

 

 

A: I don't think acquiring a mutex can throw. The rethrow has the lock guard in scope and again unlock of std::mutex should not throw. C++ reference does not mention any exceptions in relation to std::mutex.

 

+    reservation_ -= increment;

alternatively, is it possible that updateInternal() or the grow callback throw an exception for a different reason, where the reservation counter wasn't updated in the first place? Would it be safe to only have this code run for that particular exception type (the one thrown by VELOX_MEM_CAP_EXCEEDED)?

 

In velox/common/memory/MemoryUsageTracker.cpp:

 

 

A: The reservation counter is already incremented by the caller.This is true of both update and reserve, which both call this after increasing the reservation. If the reservation had not been increased there would be nothing to propagate. This same thing could be repeated inline for both callers. The point here is to keep the common case of update inlineable. This is not very important though.

 

 

 

nit: maybe some more concise like:

 if (growCallback_ && !growCallback_(type, size, *this)) {

    .. error handling block ...

  }

 

A: OK. Added qualification to GrowCallback comment that this must not throw. 

 

+      if (parent_) {

+        parent_->updateInternal(type, -size);

couldn't this trigger growCallback in the parent as well? Maybe split the functions in an updateInternal() and updateInternalUnsafe() (or similar), the ensure the latter wouldn't do the same checks and potentially trigger the callback?

—

 

A: Since the size is decremented this cannot throw.

 

Split increment and decrement of usage and made decrement noexcept to document this.

pedroerp commented on velox/common/memory/MemoryUsageTracker.cpp:
```diff
@@ -32,15 +32,31 @@ std::shared_ptr<MemoryUsageTracker> MemoryUsageTracker::create(
+    updateInternal(type_, increment);
+  } catch (std::exception& e) {
+    // Compensate for ethe increase after exceeding limit.
+    std::lock_guard<std::mutex> l(mutex_);
```
> A: I don't think acquiring a mutex can throw. The rethrow has the lock guard in scope and again unlock of std::mutex should not throw. C++ reference does not mention any exceptions in relation to std::mutex.

I just checked and the standard allows mutex::lock implementations to throw if they detect a deadlock. Let's capture the exception_ptr here and handle the exception outside of this block, for safety.

pedroerp commented on velox/common/memory/MemoryUsageTracker.cpp:
```diff
@@ -32,15 +32,31 @@ std::shared_ptr<MemoryUsageTracker> MemoryUsageTracker::create(
+  } catch (std::exception& e) {
+    // Compensate for ethe increase after exceeding limit.
+    std::lock_guard<std::mutex> l(mutex_);
+    reservation_ -= increment;
```
> A: The reservation counter is already incremented by the caller.This is true of both update and reserve, which both call this after increasing the reservation. If the reservation had not been increased there would be nothing to propagate. This same thing could be repeated inline for both callers. The point here is to keep the common case of update inlineable. This is not very important though.

Makes sense.

oerling commented:
I updated the PR. No logic inside catch.

pedroerp reviewed (approved):
Looks great, thanks for the changes and answers!
